worlds/sc2/items.py

```diff
@@ -1902,6 +1905,59 @@ def get_full_item_list():
         ItemData(812 + SC2LOTV_ITEM_ID_OFFSET, ProtossItemType.Solarite_Core, 12, SC2Race.PROTOSS, origin={"ext"}),
 }
 
+# Add keys to item table
+# Mission keys (1-999)
+mission_key_item_table = {
+    item_names._TEMPLATE_MISSION_KEY.format(mission.mission_name):
+        ItemData(mission.id + SC2_KEY_ITEM_ID_OFFSET, FactionlessItemType.Nothing, 0, SC2Race.ANY,
+                 classification=ItemClassification.progression, quantity=0, origin={"key"})
+    for mission in SC2Mission
+}
+# Numbered layout keys (1001 - 1999)
+numbered_layout_key_item_table = {
+    item_names._TEMPLATE_NUMBERED_LAYOUT_KEY.format(mission.id):
+        ItemData(mission.id + SC2_KEY_ITEM_ID_OFFSET + SC2_KEY_ITEM_SECTION_SIZE, FactionlessItemType.Nothing, 0, SC2Race.ANY,
+                 classification=ItemClassification.progression, quantity=0, origin={"key"})
+    for mission in SC2Mission
+}
+# Numbered campaign keys (2001 - 2999)
+numbered_campaign_key_item_table = {
+    item_names._TEMPLATE_NUMBERED_CAMPAIGN_KEY.format(mission.id):
+        ItemData(mission.id + SC2_KEY_ITEM_ID_OFFSET + SC2_KEY_ITEM_SECTION_SIZE * 2, FactionlessItemType.Nothing, 0, SC2Race.ANY,
+                 classification=ItemClassification.progression, quantity=0, origin={"key"})
+    for mission in SC2Mission
+}
+# Flavor keys (3000 - 3499)
+flavor_key_item_table = {
+    item_names._TEMPLATE_FLAVOR_KEY.format(name):
+        ItemData(i + SC2_KEY_ITEM_ID_OFFSET + SC2_KEY_ITEM_SECTION_SIZE * 3, FactionlessItemType.Nothing, 0, SC2Race.ANY,
+                 classification=ItemClassification.progression, quantity=0, origin={"key"})
+    for (i, name) in enumerate(item_names._flavor_key_names)
+}
+# Named layout keys (3500 - 3799)
+campaign_to_layout_names = get_used_layout_names()
+named_layout_key_item_table = {
+    item_names._TEMPLATE_NAMED_LAYOUT_KEY.format(layout_name, campaign.campaign_name):
+        ItemData(layout_start + i + SC2_KEY_ITEM_ID_OFFSET + SC2_KEY_ITEM_SECTION_SIZE * 3 + 500, FactionlessItemType.Nothing, 0, SC2Race.ANY,
+                 classification=ItemClassification.progression, quantity=0, origin={"key"})
+    for (campaign, (layout_start, layout_names)) in campaign_to_layout_names.items() for (i, layout_name) in enumerate(layout_names)
+}
+# Named campaign keys (3800 - 3999)
+campaign_names = [campaign.campaign_name for campaign in SC2Campaign if campaign != SC2Campaign.GLOBAL]
+named_campaign_key_item_table = {
+    item_names._TEMPLATE_NAMED_CAMPAIGN_KEY.format(campaign_name):
+        ItemData(i + SC2_KEY_ITEM_ID_OFFSET + SC2_KEY_ITEM_SECTION_SIZE * 3 + 800, FactionlessItemType.Nothing, 0, SC2Race.ANY,
+                 classification=ItemClassification.progression, quantity=0, origin={"key"})
+    for (i, campaign_name) in enumerate(campaign_names)
+}
```

Comment:
I'd reserve entire range for each group, like 0/1/2/3/4/5 for each group. Also add a comment that the values do start from the key offset

Reply:
I didn't feel great about reserving 6000 items when all of the existing items don't reserve that much, but if you think it's okay then I'll do that.